src/server/pkg/localcache/cache.go

```diff
@@ -0,0 +1,97 @@
+package localcache
+
+import (
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+	"sort"
+	"sync"
+)
+
+// Cache is a simple unbounded disk cache and is safe for concurrency.
+type Cache struct {
+	root string
+	keys map[string]bool
+	mu   sync.Mutex
+}
+
+// NewCache creates a new cache.
+func NewCache(root string) *Cache {
+	return &Cache{
+		root: root,
+		keys: make(map[string]bool),
+	}
+}
+
+// Put puts a key/value pair in the cache and reads the value from an io.Reader.
+func (c *Cache) Put(key string, value io.Reader) (retErr error) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	f, err := os.Create(filepath.Join(c.root, key))
+	if err != nil {
+		return err
+	}
+	defer func() {
+		if err := f.Close(); err != nil && retErr == nil {
+			retErr = err
+		}
+	}()
+	if _, err := io.Copy(f, value); err != nil {
+		return err
+	}
```

Comment:
This is probably a good place to use `io.CopyBuffer` and a buffer pool. `io.Copy` can be kinda slow by default because it uses a `4096` byte buffer by default. It can also create a lot of garbage because each call allocates a new `4096` byte buffer.

Reply:
So, when I first started implementing this, I used CopyBuffer (as well as some other optimizations to have maximum concurrency and to do as much in memory as possible), but I eventually backed off because I feel like we could be adding too much complexity with pre-mature optimizations without any data to back it up. The CopyBuffer change does not seem like much complexity, but the problem with it (and something we have encountered before) is that the buffer pooling solution we have (in grpcutil) does not defend against bursty requests. So, for example, if we were to make 1024 concurrent requests that did disk IO, we would be using 2G of memory in buffers alone. There is of course a way around this which involves some adjustments to our buffer pool or maybe making the buffers smaller, but I sort of decided that a simple solution that is guaranteed to work in every situation (coarse grained locking and no chance of using too much memory) would be a good starting point for testing things out. Since we have coarse grained locking right now, the bursty requests are not an immediate problem, but I actually think I am going to do some testing today to independently change those two variables to see which has a bigger impact, if any. Right now, it is probably not a bottleneck because we are just dealing with metadata, but it will probably matter when we starting moving the actual data around.